src/viam/sdk/components/base/server.hpp

```diff
@@ -0,0 +1,62 @@
+/// @file components/base/server.hpp
+///
+/// @brief Implements a gRPC server for the `Base` component.
+#pragma once
+
+#include <viam/api/common/v1/common.pb.h>
+#include <viam/api/component/base/v1/base.grpc.pb.h>
+
+#include <viam/sdk/resource/resource_server_base.hpp>
+#include <viam/sdk/subtype/subtype.hpp>
+
+namespace viam {
+namespace sdk {
+
+/// @class BaseServer
+/// @brief gRPC server implementation of a `Base` component.
+/// @ingroup Base
+class BaseServer : public ResourceServerBase,
+                   public viam::component::base::v1::BaseService::Service {
+   public:
+    ::grpc::Status MoveStraight(
+        ::grpc::ServerContext* context,
+        const ::viam::component::base::v1::MoveStraightRequest* request,
+        ::viam::component::base::v1::MoveStraightResponse* response) override;
+
+    ::grpc::Status Spin(::grpc::ServerContext* context,
+                        const ::viam::component::base::v1::SpinRequest* request,
+                        ::viam::component::base::v1::SpinResponse* response) override;
+
+    ::grpc::Status SetPower(::grpc::ServerContext* context,
+                            const ::viam::component::base::v1::SetPowerRequest* request,
+                            ::viam::component::base::v1::SetPowerResponse* response) override;
+
+    ::grpc::Status SetVelocity(::grpc::ServerContext* context,
+                               const ::viam::component::base::v1::SetVelocityRequest* request,
+                               ::viam::component::base::v1::SetVelocityResponse* response) override;
+
+    ::grpc::Status Stop(::grpc::ServerContext* context,
+                        const ::viam::component::base::v1::StopRequest* request,
+                        ::viam::component::base::v1::StopResponse* response) override;
+
+    ::grpc::Status IsMoving(::grpc::ServerContext* context,
+                            const ::viam::component::base::v1::IsMovingRequest* request,
+                            ::viam::component::base::v1::IsMovingResponse* response) override;
+
+    ::grpc::Status DoCommand(grpc::ServerContext* context,
+                             const viam::common::v1::DoCommandRequest* request,
+                             viam::common::v1::DoCommandResponse* response) override;
+
+    void register_server(std::shared_ptr<Server> server) override;
+
+    std::shared_ptr<SubtypeService> get_sub_svc();
+
+    BaseServer() : sub_svc(std::make_shared<SubtypeService>()){};
+    BaseServer(std::shared_ptr<SubtypeService> sub_svc) : sub_svc(sub_svc){};
+
+   private:
+    std::shared_ptr<SubtypeService> sub_svc;
```

Comment:
Given recent refactors, I think we should remove `sub_svc` as a data member and `get_sub_svc()` as a method distinct from the one in `ResourceServerBase`.

Reply:
Agreed, was waiting for that refactor to go through. Related to this: https://github.com/viamrobotics/viam-cpp-sdk/pull/89#discussion_r1180673134